README.rst
#########################
Compound Forms for Django |build|
#########################

This allows building forms that encompass other forms. Either dynamically,
by giving already-instanciated forms to one of the Proxy variants, or statically,
building a Compound form class that manages and instanciates subforms by itself.

The purpose is to ease the use of packages such as crispy-forms: a layout applied
to a Compound form can re-order fields through all sub-forms.

Just shared as is for now. Further development into a fully grown module is
intended, but not yet planned.

- Supports python 2.7, 3.3 and 3.4.
- Supports Django 1.5, 1.6 and 1.7.

It is not documented yet, though it will feel familiar if you used django
forms and formsets for a while. It is quite well tested though, with code
coverage above 90%.

If you use it, please drop me a line to let me know. Pull requests are welcome
as well.

.. |build| image:: https://travis-ci.org/spectras/django-compound-forms.svg?branch=master
    :target: https://travis-ci.org/spectras/django-compound-forms


CHANGES
v0.1.0, 2014-07-07 -- Initial release.


tests/data.py
from collections import namedtuple

#===============================================================================

NormalData = namedtuple('NormalData', 'common field_a')
NORMAL = {
    1: NormalData(common=u'common1', field_a=u'normal_a1'),
    2: NormalData(common=u'common2', field_a=u'normal_a2'),
}

NormalRelatedData = namedtuple('StandardData', 'common field_a normal')
NORMALREL = {
    1: NormalRelatedData(common=u'common1', field_a=u'normalrel_a1', normal=1),
    2: NormalRelatedData(common=u'common2', field_a=u'normalrel_a2', normal=2),
    3: NormalRelatedData(common=u'common3', field_a=u'normalrel_a3', normal=1),
    4: NormalRelatedData(common=u'common4', field_a=u'normalrel_a4', normal=2),
}

#===============================================================================

OtherData = namedtuple('OtherData', 'common field_a')
OTHER = {
    1: OtherData(common=u'common1', field_a=u'other_a1'),
    2: OtherData(common=u'common2', field_a=u'other_a2'),
}

OtherRelatedData = namedtuple('StandardData', 'common field_a other')
OTHERREL = {
    1: OtherRelatedData(common=u'common1', field_a=u'otherrel_a1', other=2),
    2: OtherRelatedData(common=u'common2', field_a=u'otherrel_a2', other=1),
    3: OtherRelatedData(common=u'common3', field_a=u'otherrel_a3', other=2),
    4: OtherRelatedData(common=u'common4', field_a=u'otherrel_a4', other=1),
}


tests/forms.py
from django.forms import CharField
from collections import OrderedDict
from compound_forms.forms import (MergingProxyForm, MergingCompoundModelForm,
                                  compoundform_factory)

from app.models import Normal, Other
from app.forms import NormalForm, OtherForm
from .data import NORMAL, OTHER
from .fixtures import NormalFixture, OtherFixture
from .formdata import FormData
from .utils import TestCase


class BasicProxyFormTest(NormalFixture, OtherFixture, TestCase):
    normal_count = 1
    other_count = 1

    """ Proxy forms act as proxy towards already existing forms """
    def _get_form(self, normal, other, **kwargs):
        ''' Helper function to build the form '''
        defaults = {
            'forms': OrderedDict((('normal', normal), ('other', other),)),
        }
        defaults.update(kwargs)
        return MergingProxyForm(**defaults)

    def test_basic_proxy_create(self):
        form = self._get_form(NormalForm(), OtherForm())
        # Test the forms are set on the proxy form, and regular form interface works
        self.assertEqual(tuple(form.forms.keys()), ('normal', 'other'))
        self.assertCountEqual(form.fields,
                              ('normal.common', 'normal.field_a',
                               'other.common', 'other.field_a'))
        self.assertCountEqual((field.name for field in form),
                              ('common', 'field_a',
                               'common', 'field_a'))

        # Test the field are actually the same
        self.assertIs(form.fields['normal.common'], form.forms['normal'].fields['common'])
        self.assertIs(form.fields['normal.field_a'], form.forms['normal'].fields['field_a'])
        self.assertIs(form.fields['other.common'], form.forms['other'].fields['common'])
        self.assertIs(form.fields['other.field_a'], form.forms['other'].fields['field_a'])

    def test_basic_proxy_initial(self):
        normal = Normal.objects.get(pk=self.normal_id[1])
        other = Other.objects.get(pk=self.other_id[1])
        form = self._get_form(NormalForm(instance=normal, prefix='normal'),
                              OtherForm(instance=other, prefix='other'))

        self.assertEqual(form['normal.common'].value(), NORMAL[1].common)
        self.assertEqual(form['normal.field_a'].value(), NORMAL[1].field_a)
        self.assertEqual(form['other.common'].value(), OTHER[1].common)
        self.assertEqual(form['other.field_a'].value(), OTHER[1].field_a)
        self.assertEqual(form.forms['normal']['common'].value(), NORMAL[1].common)
        self.assertEqual(form.forms['normal']['field_a'].value(), NORMAL[1].field_a)
        self.assertEqual(form.forms['other']['common'].value(), OTHER[1].common)
        self.assertEqual(form.forms['other']['field_a'].value(), OTHER[1].field_a)

    def test_basic_proxy_validate_correct(self):
        normal = Normal.objects.get(pk=self.normal_id[1])
        other = Other.objects.get(pk=self.other_id[1])
        form = self._get_form(NormalForm(instance=normal, prefix='normal'),
                              OtherForm(instance=other, prefix='other'))

        # Get and alter html form data
        data = FormData(form)
        data.set_form_field(form, 'normal.common', 'updated_nc')
        data.set_form_field(form, 'other.field_a', 'updated_ofa')

        # Feed it to a form and validate it
        form = self._get_form(NormalForm(instance=normal, prefix='normal', data=data),
                              OtherForm(instance=other, prefix='other', data=data),
                              data=data)

        self.assertTrue(form.is_valid())
        self.assertTrue(form.forms['normal'].is_valid())
        self.assertTrue(form.forms['other'].is_valid())
        self.assertEqual(len(form.errors), 0)

        # Check cleaned data
        self.assertEqual(form.cleaned_data['normal.common'], 'updated_nc')
        self.assertEqual(form.cleaned_data['normal.field_a'], NORMAL[1].field_a)
        self.assertEqual(form.cleaned_data['other.common'], OTHER[1].common)
        self.assertEqual(form.cleaned_data['other.field_a'], 'updated_ofa')

        # Check changed data
        self.assertTrue(form.has_changed())
        self.assertCountEqual(form.changed_data, ('normal.common', 'other.field_a'))

        # Check the forms can be saved
        form.forms['normal'].save()
        form.forms['other'].save()
        self.assertEqual(normal.pk, self.normal_id[1])
        self.assertEqual(normal.common, 'updated_nc')
        self.assertEqual(normal.field_a, NORMAL[1].field_a)
        self.assertEqual(other.pk, self.other_id[1])
        self.assertEqual(other.common, OTHER[1].common)
        self.assertEqual(other.field_a, 'updated_ofa')

    def test_basic_proxy_validate_errors(self):
        normal = Normal.objects.get(pk=self.normal_id[1])
        other = Other.objects.get(pk=self.other_id[1])
        form = self._get_form(NormalForm(instance=normal, prefix='normal'),
                              OtherForm(instance=other, prefix='other'))

        # Get and alter html form data
        data = FormData(form)
        data.set_form_field(form, 'normal.common', '')
        data.set_form_field(form, 'other.field_a', 'updated_ofa')

        # Feed it to a form and validate it
        form = self._get_form(NormalForm(instance=normal, prefix='normal', data=data),
                              OtherForm(instance=other, prefix='other', data=data),
                              data=data)

        self.assertFalse(form.is_valid())
        self.assertFalse(form.forms['normal'].is_valid())
        self.assertTrue(form.forms['other'].is_valid())
        self.assertCountEqual(form.errors, ('normal.common',))

        # Check changed data
        self.assertTrue(form.has_changed())
        self.assertCountEqual(form.changed_data, ('normal.common', 'other.field_a'))


class BasicCompoundFormTest(NormalFixture, OtherFixture, TestCase):
    """ Compound forms are Proxy form that create and manage the proxied forms """
    normal_count = 1
    other_count = 1

    def _get_form(self, **kwargs):
        form_klass = compoundform_factory(
            OrderedDict((('normal', NormalForm), ('other', OtherForm))),
            base=MergingCompoundModelForm,
        )
        return form_klass(**kwargs)

    def test_basic_compound_create(self):
        """ Create an unbound compound form to check fields are mapped correctly """
        form = self._get_form()
        self.assertEqual(tuple(form.forms.keys()), ('normal', 'other'))
        self.assertCountEqual(form.fields,
                              ('normal.common', 'normal.field_a',
                               'other.common', 'other.field_a'))
        self.assertCountEqual((field.name for field in form),
                              ('common', 'field_a',
                               'common', 'field_a'))

        # Test the field are actually the same
        self.assertIs(form.fields['normal.common'], form.forms['normal'].fields['common'])
        self.assertIs(form.fields['normal.field_a'], form.forms['normal'].fields['field_a'])
        self.assertIs(form.fields['other.common'], form.forms['other'].fields['common'])
        self.assertIs(form.fields['other.field_a'], form.forms['other'].fields['field_a'])

        # Test sub-forms have prefixes
        self.assertEqual(form.forms['normal'].prefix, 'normal')
        self.assertEqual(form.forms['other'].prefix, 'other')

    def test_basic_compound_initial(self):
        """ Initialize the form with an instance on Normal and no Other """
        normal = Normal.objects.get(pk=self.normal_id[1])
        form = self._get_form(instances={'normal': normal})

        self.assertEqual(form['normal.common'].value(), NORMAL[1].common)
        self.assertEqual(form['normal.field_a'].value(), NORMAL[1].field_a)
        self.assertEqual(form['other.common'].value(), None)
        self.assertEqual(form['other.field_a'].value(), None)
        self.assertEqual(form.forms['normal']['common'].value(), NORMAL[1].common)
        self.assertEqual(form.forms['normal']['field_a'].value(), NORMAL[1].field_a)
        self.assertEqual(form.forms['other']['common'].value(), None)
        self.assertEqual(form.forms['other']['field_a'].value(), None)

    def test_basic_compound_validate(self):
        """ Update a Normal and create a new Other in one submission """
        normal = Normal.objects.get(pk=self.normal_id[1])
        form = self._get_form(instances={'normal': normal})

        # Get and alter html form data
        data = FormData(form)
        data.set_form_field(form, 'normal.common', 'updated_nc')
        data.set_form_field(form, 'other.common', 'created_oc')
        data.set_form_field(form, 'other.field_a', 'created_ofa')

        # Feed it to a form and validate it
        form = self._get_form(instances={'normal': normal}, data=data)

        self.assertTrue(form.is_valid())
        self.assertTrue(form.forms['normal'].is_valid())
        self.assertTrue(form.forms['other'].is_valid())
        self.assertEqual(len(form.errors), 0)

        # Check the forms can be saved
        result = form.save()
        self.assertIs(result['normal'], normal)
        self.assertEqual(normal.pk, self.normal_id[1])
        self.assertEqual(normal.common, 'updated_nc')
        self.assertEqual(normal.field_a, NORMAL[1].field_a)
        self.assertNotIn(result['other'].pk, (None,) + tuple(self.other_id.values()))
        self.assertEqual(result['other'].common, 'created_oc')
        self.assertEqual(result['other'].field_a, 'created_ofa')


class LinkedCompoundFormTest(NormalFixture, OtherFixture, TestCase):
    """ Compound forms with linked fields """
    normal_count = 1
    other_count = 1

    def _get_form(self, **kwargs):
        form = compoundform_factory(
            OrderedDict((('normal', NormalForm), ('other', OtherForm))),
            linked_fields=OrderedDict((('common', CharField(max_length=255, required=False)),)),
            base=MergingCompoundModelForm,
        )
        return form(**kwargs)

    def test_linked_compound_create(self):
        """ Create an unbound compound form to check fields are mapped correctly """
        form = self._get_form()
        self.assertEqual(tuple(form.forms.keys()), ('normal', 'other'))
        # Common field should shadow subform field of same name
        self.assertCountEqual(form.fields, ('common', 'normal.field_a', 'other.field_a'))
        self.assertCountEqual((field.name for field in form),
                              ('common', 'field_a', 'field_a'))

    def test_linked_compound_initial(self):
        # Test with matching instances
        assert NORMAL[1].common == OTHER[1].common, 'common field must match for same id'
        normal = Normal.objects.get(pk=self.normal_id[1])
        other = Other.objects.get(pk=self.other_id[1])

        # With matching linked fields, everything should be fine
        form = self._get_form(instances={'normal': normal, 'other': other})
        self.assertEqual(form['common'].value(), NORMAL[1].common)

        # With mismatched linked fields, ValueError should be raised
        with self.assertRaises(ValueError):
            form = self._get_form(instances={'normal': normal})

        # Specifiying an initial value manually should bypass pulling
        form = self._get_form(instances={'normal': normal},
                              initial={'common': 'common_force'})
        self.assertEqual(form['common'].value(), 'common_force')

        # Linked field should shadow subform's fields, including initial setting
        with self.assertRaises(ValueError):
            form = self._get_form(instances={'normal': normal},
                                  initial={'other.common': NORMAL[1].common})

    def test_linked_compound_validate_correct(self):
        """ Update a Normal and create a new Other in one submission """
        normal = Normal.objects.get(pk=self.normal_id[1])
        other = Other.objects.get(pk=self.other_id[1])
        form = self._get_form(instances={'normal': normal, 'other': other})

        # Get and alter html form data
        data = FormData(form)
        data.set_form_field(form, 'common', 'updated_common')
        data.set_form_field(form, 'other.field_a', 'updated_ofa')

        # Feed it to a form and validate it
        form = self._get_form(instances={'normal': normal, 'other': other}, data=data)

        self.assertTrue(form.is_valid())
        self.assertTrue(form.forms['normal'].is_valid())
        self.assertTrue(form.forms['other'].is_valid())
        self.assertEqual(len(form.errors), 0)
        self.assertCountEqual(form.cleaned_data,
                              ('common', 'normal.field_a', 'other.field_a'))

        # Check changed data
        self.assertTrue(form.has_changed())
        self.assertCountEqual(form.changed_data, ('common', 'other.field_a'))

        # Check the forms can be saved
        result = form.save()
        self.assertIs(result['normal'], normal)
        self.assertIs(result['other'], other)
        self.assertEqual(normal.pk, self.normal_id[1])
        self.assertEqual(normal.common, 'updated_common')
        self.assertEqual(normal.field_a, NORMAL[1].field_a)
        self.assertEqual(other.pk, self.other_id[1])
        self.assertEqual(other.common, 'updated_common')
        self.assertEqual(other.field_a, 'updated_ofa')

    def test_linked_compound_validate_incorrect(self):
        normal = Normal.objects.get(pk=self.normal_id[1])
        other = Other.objects.get(pk=self.other_id[1])
        form = self._get_form(instances={'normal': normal, 'other': other})

        # Get and alter html form data
        data = FormData(form)
        data.set_form_field(form, 'common', '')

        # Feed it to a form and validate it
        form = self._get_form(instances={'normal': normal, 'other': other}, data=data)

        self.assertFalse(form.is_valid())
        self.assertFalse(form.forms['normal'].is_valid())
        self.assertFalse(form.forms['other'].is_valid())
        self.assertCountEqual(form.errors, ('common',))

        # Check changed data
        self.assertTrue(form.has_changed())
        self.assertCountEqual(form.changed_data, ('common',))


tests/tests.py
import django
if django.VERSION < (1, 6):
    from .fixtures import FixtureTests
    from .forms import (BasicProxyFormTest, BasicCompoundFormTest,
                        LinkedCompoundFormTest)
    from .formsets import (ProxyFormSetTests, CompoundInlineFormSetTests)

compound_forms/forms.py
from django.forms import Form, ModelForm
from django.forms.forms import BaseForm, NON_FIELD_ERRORS
from django.utils.functional import cached_property
from collections import OrderedDict
import copy

##############################################################################

class SubFormsProxyMixin(BaseForm):
    """ Base form that handles sub-forms with optional linked fields """
    linked_fields = OrderedDict()

    def __init__(self, *args, **kwargs):
        pull_linked_fields = kwargs.pop('pull_linked_fields', True)
        super(SubFormsProxyMixin, self).__init__(*args, **kwargs)
        for name, field in self.linked_fields.items():
            if field is not None:
                # as it can be modified, each form must have it own field instance
                self.fields[name] = copy.deepcopy(field)
        if pull_linked_fields:
            self.pull_linked_fields()

    def pull_linked_fields(self):
        """ Pull initial values from sub-forms (and checks they are identical) """
        if not self.is_bound:
            for name in self.linked_fields.keys():
                if name in self.initial: # if an initial value is explicitly
                    continue             # specified, simply use it
                if name not in self.fields:  # field will be added by add_field
                    continue                 # on some formset, just skip pulling it
                initials = tuple(
                    form.initial.get(name, form.fields[name].initial)
                    for form in self.forms.values()
                    if name in form.fields
                )
                if initials.count(initials[0]) != len(initials):
                    raise ValueError('Initial sub-form values differ for '
                                     'linked field %s: %r' % (name, initials))
                self.initial[name] = initials[0]

    def push_linked_fields(self):
        """ Push raw field data to sub-forms and let them do the cleaning later """
        if self.is_bound:
            for form in self.forms.values():
                form.data = form.data.copy() # we need a mutable copy
                form.data.update(dict(
                    (form.add_prefix(name), self._raw_value(name))
                    for name in self.linked_fields.keys()
                    if name in form.fields
                ))

    def full_clean(self):
        """ Hook field data pushing before form cleaning kicks in """
        if self.is_bound:
            self.push_linked_fields()
        super(SubFormsProxyMixin, self).full_clean()

    @property
    def media(self):
        return (super(SubFormsProxyMixin, self).media +
                sum(form.media for form in self.forms.values()))

    def is_multipart(self):
        return (super(SubFormsProxyMixin, self).is_multipart() or
                any(form.is_multipart() for form in self.forms.values()))

##############################################################################

class SubFormsBuildMixin(BaseForm):
    form_classes = OrderedDict()

    @cached_property
    def forms(self):
        """ Dict of name: form """
        return OrderedDict((name, self._construct_form(name)) for name in self.form_classes.keys())

    def _construct_form(self, name, **kwargs):
        klass = self.form_classes[name]
        defaults = {
            'auto_id': self.auto_id,
            'prefix': self.add_prefix(name),
            'error_class': self.error_class,
        }
        if self.is_bound:
            defaults['data'] = self.data
            defaults['files'] = self.files
        if self.initial and not 'initial' in kwargs:
            try:
                defaults['initial'] = self.initial[name]
            except KeyError:
                pass
        defaults.update(kwargs)
        return klass(**defaults)

##############################################################################

class ModelSubFormsMixin(BaseForm):
    def __init__(self, *args, **kwargs):
        self.instances = kwargs.pop('instances', {})
        super(ModelSubFormsMixin, self).__init__(*args, **kwargs)

    def _construct_form(self, name, **kwargs):
        defaults = {
            'instance': self.instances.get(name),
        }
        defaults.update(kwargs)
        return super(ModelSubFormsMixin, self)._construct_form(name, **defaults)

    def save(self, only=None, **kwargs):
        keys = self.forms.keys() if only is None else only
        return OrderedDict((name, self._save_form(name, **kwargs)) for name in keys)

    def _save_form(self, name, **kwargs):
        return self.forms[name].save(**kwargs)

##############################################################################

class MergingFormMixin(BaseForm):
    """ A BaseCompoundForm that allows access to subforms through field aliases """
    def __init__(self, *args, **kwargs):
        super(MergingFormMixin, self).__init__(*args, **kwargs)
        self.field_form = {}
        self._make_field_aliases()

    def _make_field_aliases(self):
        for form_name, form in self.forms.items():
            for field_name, field in form.fields.items():
                if field_name in self.linked_fields:
                    continue
                alias = self._get_alias(form_name, field_name)
                self.fields[alias] = field
                self.field_form[alias] = (form, field_name)

    def _construct_form(self, name, **kwargs):
        defaults = {}
        if 'initial' in kwargs: # extract initial for this form
            initial = {}
            for key, value in kwargs['initial'].items():
                try:
                    form_name, field_name = key.split('.')
                except ValueError:
                    continue
                if form_name == name and field_name not in self.linked_fields:
                    initial[field_name] = value
            defaults['initial'] = initial
        defaults.update(kwargs)
        return super(MergingFormMixin, self)._construct_form(name, **defaults)

    def __getitem__(self, name):
        """ Retrieve the field by name, merging own field with content form's """
        if name in self.field_form:
            form, field_name = self.field_form[name]
            return form.__getitem__(field_name)
        return super(MergingFormMixin, self).__getitem__(name)

    def _clean_fields(self):
        """ Omit merged fields from _clean_fields, as they would be seen as all empty """
        fields = self.fields
        self.fields = dict((k, v) for k, v in fields.items() if not k in self.field_form)
        super(MergingFormMixin, self)._clean_fields()
        self.fields = fields

    def _clean_form(self):
        """ Merge in subform errors and cleaned_data under their alias names """
        super(MergingFormMixin, self)._clean_form()
        for form_name, form in self.forms.items():
            for field_name, errors in form.errors.items():
                if field_name != NON_FIELD_ERRORS:
                    field_name = self._get_alias(form_name, field_name)
                self._errors.setdefault(field_name, set()).update(errors)
            self.cleaned_data.update((self._get_alias(form_name, name), data)
                                     for name, data in form.cleaned_data.items()
                                     if name not in self.linked_fields)

    @property
    def changed_data(self):
        """ Merge in subform's changed_data """
        if self._changed_data is None:
            ret = set(super(MergingFormMixin, self).changed_data)
            for form_name, form in self.forms.items():
                ret.update(self._get_alias(form_name, name) for name in form.changed_data)
            self._changed_data = tuple(ret)
        return self._changed_data

    def _get_alias(self, form_name, field_name):
        """ Compute the alias for a given form and field names """
        if field_name in self.linked_fields:
            return field_name
        return '%s.%s' % (form_name, field_name)

##############################################################################

class BaseProxyForm(SubFormsProxyMixin):
    """ Compound form that proxies fields to its subforms """
    def __init__(self, *args, **kwargs):
        self.forms = kwargs.pop('forms')
        self.linked_fields = self.linked_fields.copy()
        self.linked_fields.update(kwargs.pop('linked_fields', {}))
        super(BaseProxyForm, self).__init__(*args, **kwargs)

class BaseMergingProxyForm(MergingFormMixin, BaseProxyForm):
    pass

class ProxyForm(Form, BaseProxyForm):
    pass

class MergingProxyForm(Form, BaseMergingProxyForm):
    pass


class BaseCompoundForm(SubFormsBuildMixin, SubFormsProxyMixin):
    """ Compound form that builds subforms and proxies fields to them """
    pass

class BaseMergingCompoundForm(MergingFormMixin, BaseCompoundForm):
    pass

class CompoundForm(Form, BaseCompoundForm):
    pass

class CompoundModelForm(Form, ModelSubFormsMixin, BaseCompoundForm):
    pass

class MergingCompoundForm(Form, BaseMergingCompoundForm):
    pass

class MergingCompoundModelForm(Form, ModelSubFormsMixin, BaseMergingCompoundForm):
    pass

def compoundform_factory(forms, base=MergingCompoundForm, linked_fields=None):
    attrs = {
        'form_classes': forms,
    }
    if linked_fields is not None:
        attrs['linked_fields'] = linked_fields
    return type(base.__name__, (base,), attrs)


test_project/settings.py
# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
import os
BASE_DIR = os.path.dirname(os.path.dirname(__file__))

SECRET_KEY = 'b6s4u*8q7&gh1(d04gasmx!^8y_zw#4z*2t90-tgm29*0u*9tb'
DEBUG = True
TEMPLATE_DEBUG = True

# Application definition
INSTALLED_APPS = (
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'app',
    'tests',
)

MIDDLEWARE_CLASSES = (
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
)

ROOT_URLCONF = 'project.urls'

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': os.path.join(BASE_DIR, 'db.sqlite3'),
    }
}

LANGUAGE_CODE = 'en'
TIME_ZONE = 'UTC'
USE_I18N = True
USE_L10N = True
USE_TZ = True


compound_forms/formsets.py
from django.core.exceptions import ValidationError
from django.forms.formsets import (BaseFormSet,
                                   ORDERING_FIELD_NAME, DELETION_FIELD_NAME)
from django.utils.functional import cached_property
from collections import OrderedDict

from .forms import MergingProxyForm

##############################################################################

class InvalidFormsetsError(ValueError):
    pass

##############################################################################

class SubFormSetsBuildMixin(BaseFormSet):
    formset_classes = OrderedDict()

    @cached_property
    def formsets(self):
        return OrderedDict((name, self._construct_formset(name))
                           for name in self.formset_classes.keys())

    def _construct_formset(self, name, **kwargs):
        klass = self.formset_classes[name]
        defaults = {
            'auto_id': self.auto_id,
            'prefix': self.add_prefix(name),
            'error_class': self.error_class,
        }
        if self.is_bound:
            defaults['data'] = self.data
            defaults['files'] = self.files
        if self.initial:
            defaults['initial'] = self.initial
        defaults.update(kwargs)
        return klass(**defaults)

##############################################################################

class InlineSubFormSetsMixin(SubFormSetsBuildMixin):
    def __init__(self, *args, **kwargs):
        self.instances = kwargs.pop('instances', {})
        super(InlineSubFormSetsMixin, self).__init__(*args, **kwargs)
        for name in self.instances:
            if not name in self.formsets:
                raise KeyError('Instance for unkown formset %r', name)

    def _construct_formset(self, name, **kwargs):
        defaults = {
            'instance': self.instances.get(name),
        }
        defaults.update(kwargs)
        return super(InlineSubFormSetsMixin, self)._construct_formset(name, **defaults)

    def save(self, only=None, **kwargs):
        keys = self.formsets.keys() if only is None else only
        return OrderedDict((name, self._save_formset(name, **kwargs)) for name in keys)

    def _save_formset(self, name, **kwargs):
        return self.formsets[name].save(**kwargs)

##############################################################################

class SubFormSetsProxyMixin(BaseFormSet):
    form = MergingProxyForm
    formset_group_fields = OrderedDict()
    validate_max = False

    @property
    def management_form(self):
        forms = dict((name, formset.management_form) for name, formset in self.formsets.items())
        kwargs = {
            'forms': forms,
        }
        if self.is_bound:
            kwargs['data'] = self.data
            kwargs['files'] = self.files
        form = MergingProxyForm(**kwargs)
        if self.is_bound and not form.is_valid():
            raise ValidationError('Invalid management form')
        return form

    def _construct_forms(self):
        """ On django < 1.6, this method was used instaed of a cached property """
        pass

    @cached_property
    def forms(self):
        #TODO: peek at initial data to guess the number beforehand and alter
        # subformset's management form TOTAL/extra
        groups, extras = OrderedDict(), []
        # Fill groups with forms from formsets
        first = True
        for formset_name, formset in self.formsets.items():
            # Group initial forms by key (generated from fields in formset_group_fields)
            for form in formset.initial_forms:
                key = tuple(form.initial.get(field, form.fields[field].initial)
                            for field in self.formset_group_fields.keys())
                if first:
                    groups[key] = {formset_name: form}
                else:   # ensure a mismatch key raises an exception
                    groups[key][formset_name] = form

            # Check that formset grouping is correct
            if not first:
                if formset.initial_form_count() != len(groups):
                    raise InvalidFormsetsError(
                        'formsets do not have the same number of initial form groups: %d != %d' %
                        (formset.initial_form_count(), len(groups))
                    )
                if len(formset.extra_forms) != len(extras):
                    raise InvalidFormsetsError(
                        'formsets do not have the same number of extra forms: %d != %d' %
                        (len(formset.extra_forms), len(extras))
                    )

            # Simply group extra forms by their index
            for index, form in enumerate(formset.extra_forms):
                if first:
                    extras.append({formset_name: form})
                else:
                    extras[index][formset_name] = form

            first = False

        linked_fields = self.formset_group_fields.copy()
        if self.can_order:
            linked_fields[ORDERING_FIELD_NAME] = None
        if self.can_delete:
            linked_fields[DELETION_FIELD_NAME] = None

        forms = (tuple(self._construct_form(i, forms=group, linked_fields=linked_fields)
                       for i, group in enumerate(groups.values()))
                 +
                 tuple(self._construct_form(i, forms=group, linked_fields=linked_fields)
                       for i, group in enumerate(extras, len(groups))))
        return forms

    def initial_form_count(self):
        count = next(iter(self.formsets.values())).initial_form_count()
        if any(formset.initial_form_count() != count for formset in self.formsets.values()):
            raise InvalidFormsetsError('initial_form_count()s differ amongst sub-formsets')
        return count

    def total_form_count(self):
        count = next(iter(self.formsets.values())).total_form_count()
        if any(formset.total_form_count() != count for formset in self.formsets.values()):
            raise InvalidFormsetsError('total_form_count()s differ amongst sub-formsets')
        return count

    def full_clean(self):
        self._errors = []
        self._non_form_errors = self.error_class()
        if not self.is_bound:
            return

        for i in range(0, self.total_form_count()):
            self.forms[i].push_linked_fields()

        unique_non_form_errors = set()
        for formset in self.formsets.values():
            unique_non_form_errors.update(formset.non_form_errors())
            # do not add form errors as we will get them right after
        self._non_form_errors.extend(unique_non_form_errors)

        for form in self.forms:
            self._errors.append(form.errors)
        try:
            self.clean()
        except ValidationError as e:
            self._non_form_errors = self.error_class(e.messages)

    @property
    def min_num(self):
        """ Django >= 1.7 """
        return max(formset.min_num for formset in self.formsets.values())

    @property
    def can_order(self):
        return all(formset.can_order for formset in self.formsets.values())

    @property
    def can_delete(self):
        return all(formset.can_delete for formset in self.formsets.values())

    @property
    def empty_form(self):
        form_list = tuple(formset.empty_form for formset in self.formsets.values())
        form = self.form(forms=form_list)
        self.add_fields(form, None)
        return form

##############################################################################

class ProxyFormSet(SubFormSetsProxyMixin, BaseFormSet):
    def __init__(self, *args, **kwargs):
        self.formsets = kwargs.pop('formsets')
        self.formset_group_fields = self.formset_group_fields.copy()
        self.formset_group_fields.update(kwargs.pop('formset_group_fields', {}))
        super(ProxyFormSet, self).__init__(*args, **kwargs)

class CompoundFormSet(SubFormSetsBuildMixin, SubFormSetsProxyMixin, BaseFormSet):
    pass

class CompoundInlineFormSet(InlineSubFormSetsMixin, SubFormSetsProxyMixin, BaseFormSet):
    pass

def compoundformset_factory(formsets, base=CompoundFormSet, formset_group_fields=None):
    attrs = {
        'formset_classes': formsets,
    }
    if formset_group_fields is not None:
        attrs['formset_group_fields'] = formset_group_fields
    return type(base.__name__, (base,), attrs)
